apps/hash-api/src/graph/knowledge/primitive/entity/after-update-entity-hooks.ts

```diff
@@ -1,205 +1,24 @@
-          triggeredByUser.accountId !== addedMentionedUser.accountId,
-      )
-      .map(async (addedMentionedUser) => {
-        const { view: mentionedUserCanViewPage } =
-          await checkPermissionsOnEntity(
-            context,
-            { actorId: addedMentionedUser.accountId },
-            { entity: occurredInEntity.entity },
-          );
-
-        if (!mentionedUserCanViewPage) {
-          return;
-        }
-
-        const existingNotification = await getMentionNotification(
-          context,
-          /** @todo: use authentication of machine user instead */
-          { actorId: addedMentionedUser.accountId },
-          {
-            recipient: addedMentionedUser,
-            triggeredByUser,
-            occurredInEntity,
-            occurredInComment,
-            occurredInBlock,
-            occurredInText: text,
-          },
-        );
-
-        if (!existingNotification) {
-          await createMentionNotification(
-            context,
-            /** @todo: use authentication of machine user instead */
-            { actorId: addedMentionedUser.accountId },
-            {
-              ownedById: addedMentionedUser.accountId as OwnedById,
-              occurredInEntity,
-              occurredInBlock,
-              occurredInComment,
-              occurredInText: text,
-              triggeredByUser,
-            },
-          );
-        }
-      }),
-  ]);
-};
-
-const userAfterUpdateHookCallback: UpdateEntityHookCallback = async ({
-  entity,
-}) => {
-  if (isProdEnv) {
-    const {
-      email: emails,
-      shortname,
-      preferredName,
-    } = simplifyProperties(entity.properties as UserProperties);
-
-    /**
-     * @todo: when we allow users to have more than one email, come up with
-     * a better way of determining which to use for mailchimp.
-     */
-    const [email] = emails;
-
-    await createOrUpdateMailchimpUser({
-      email,
-      shortname,
-      preferredName,
-    });
-  }
-};
+  entityTypesToParseTextFrom,
+  parseTextFromFileAfterUpdateEntityHookCallback,
+} from "./after-update-entity-hooks/file-document-after-update-entity-hook-callback";
+import { textAfterUpdateEntityHookCallback } from "./after-update-entity-hooks/text-after-update-entity-hook-callback";
+import { userAfterUpdateEntityHookCallback } from "./after-update-entity-hooks/user-after-update-entity-hook";
+import { UpdateEntityHook } from "./update-entity-hooks";
 
 export const afterUpdateEntityHooks: UpdateEntityHook[] = [
   {
     entityTypeId: systemEntityTypes.text.entityTypeId,
-    callback: textEntityUpdateHookCallback,
+    callback: textAfterUpdateEntityHookCallback,
   },
   {
     entityTypeId: systemEntityTypes.user.entityTypeId,
-    callback: userAfterUpdateHookCallback,
+    callback: userAfterUpdateEntityHookCallback,
   },
+  ...entityTypesToParseTextFrom.map((entityTypeId) => ({
+    entityTypeId,
+    callback: parseTextFromFileAfterUpdateEntityHookCallback,
+  })),
```

Comment:
I should've thought of this sooner, but how this does actually get reliably triggered? The user uploads the file after the entity is created, and the file is not mutated after the upload is complete.

Reply:
I've addressed this via a `uploadCompletedAt` property, as discussed internally. This involved several commits, I recommend looking at the diff since your review to see the changes.